### PR TITLE
Add abstractions for Razor IR

### PR DESCRIPTION
**Pull request by rynowak**
This is an API skeleton for the IR data model that we'll be using as a
spiritual continuation of the 'chunks' API. Currently missing a lot of
detail which will be filled in as needed.

### CODE REVIEW

File: src/Microsoft.AspNetCore.Razor.Evolution/Intermediate/IRNodeVisitor.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+namespace Microsoft.AspNetCore.Razor.Evolution.Intermediate
+{
+    public abstract class IRNodeVisitor 
+    {
+        public virtual void Visit(IRNode node)
+        {
+            node.Accept(this);
+        }
+
+        public virtual void VisitDefault(IRNode node)
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Evolution/Intermediate/IRNodeVisitor.cs]: lol Roslyn

**Comment by NTaylorMullen** [comment]
Also, not sure I'm 100% convinced on the names of IRX because we typically prefix interfaces with "I" in the asp.net repos.

File: src/Microsoft.AspNetCore.Razor.Evolution/Intermediate/IRNode.cs [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+using System.Collections.Generic;
+
+namespace Microsoft.AspNetCore.Razor.Evolution.Intermediate
+{
+    public abstract class IRNode
+    {
+        internal static readonly IRNode[] EmptyArray = new IRNode[0];
+
+        public abstract IList<IRNode> Children { get; }
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Evolution/Intermediate/IRNode.cs]: Part of me thinks we should have two classes. Parent and leaf nodes.

**Comment by rynowak** [comment]
If you have a better suggestion we do that.. I don't want to just block on naming right now though

**Comment by NTaylorMullen** [comment]
> If you have a better suggestion we do that.. I don't want to just block on naming right now though

Agree which is why I signed off already 😄

**Pull request merged**